src/coord/src/persistcfg.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Materialize-specific persistence configuration.
+
+use std::path::PathBuf;
+
+use anyhow::anyhow;
+use serde::Serialize;
+
+use expr::GlobalId;
+use persist::file::{FileBlob, FileBuffer};
+use persist::indexed::runtime::{self, RuntimeClient, StreamWriteHandle};
+
+/// Configuration of the persistence runtime and features.
+#[derive(Clone, Debug)]
+pub struct PersistConfig {
+    /// A directory under which un-indexed WAL-like writes are quickly stored.
+    pub buffer_path: PathBuf,
+    /// A directory under which larger batches of indexed data are stored. This
+    /// will eventually be S3 for Cloud.
+    pub blob_path: PathBuf,
+    /// Whether to persist user tables. This is extremely experimental and
+    /// should not even be tried by users. It's initially here for end-to-end
+    /// testing.
+    pub user_table_enabled: bool,
+    /// Information stored in the "lock" files created by the buffer and blob to
+    /// ensure that they are exclusive writers to those locations. This should
+    /// contain whatever information might be useful to investigating an
+    /// unexpected lock file (e.g. hostname and materialize version of the
+    /// creating process).
+    pub lock_info: String,
+}
+
+impl PersistConfig {
+    pub fn disabled() -> Self {
+        PersistConfig {
+            buffer_path: Default::default(),
+            blob_path: Default::default(),
+            user_table_enabled: false,
+            lock_info: Default::default(),
+        }
+    }
+
+    /// Initializes the persistence runtime and returns a clone-able handle for
+    /// interacting with it. Returns None and does not start the runtime if all
+    /// persistence features are disabled.
+    pub fn init(&self) -> Result<PersisterWithConfig, anyhow::Error> {
+        let persister = if self.user_table_enabled {
```

Comment:
haven't read through the whole thing yet but this condition on `user_table_enabled` is confusing to me -- it seems like if `user_table_enabled` is false then all tables (user and system) will not be persisted (this could be a misread). Not sure if that is the current intent and perhaps the flag should be renamed or if the intent is to always persist system tables and only conditionally persist user tables

Reply:
System tables are unconditionally disabled elsewhere right now (the persisted field is false for the definition of each). When we turn them on by default, we'll have to add an opt-out escape hatch flag, so I think this is fine for now?